sycl/doc/extensions/proposed/sycl_ext_oneapi_p2p.asciidoc

```diff
@@ -0,0 +1,157 @@
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+This extension adds support for mechanisms to query and enable support for
+direct memory access between peer devices in a system.
+In particular, this allows one device to directly access USM Device allocations
+for a peer device. Peer to peer capabilities are useful as they can provide
+access to a peer device's memory inside a compute kernel and optimized memory
+copies between peer devices.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_P2P` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+
+=== Peer to Peer (P2P) Memory Access APIs
+
+This extension adds support for mechanisms to query and enable support for
+direct memory access between peer devices in a system.
+In particular, this allows one device to directly access USM Device
+allocations for a peer device in the same context.
+Peer to peer capabilities are useful as they can provide access to a peer
+device's memory inside a compute kernel and also optimized memory copies between
+peer devices.
+
+This extension adds four new member functions to the device class, as described
+below.
+
+[source,c++]
+----
+namespace sycl   {
+namespace ext    {
+namespace oneapi {
+enum class p2p_property {
+  access_supported,
+  atomics_supported
+};
+} // namespace oneapi
+} // namespace ext
+
+class device {
+public:
+  bool ext_oneapi_can_access_peer(const device &peer);
+  void ext_oneapi_enable_peer_access(const device &peer);
+  void ext_oneapi_disable_peer_access(const device &peer);
+  int ext_oneapi_get_p2p_property(const device &peer, ext::oneapi::p2p_property value);
+};
+
+} // namespace sycl
+----
+
+The semantics of the new functions are:
+
+|===
+| Member Function | Description
+
+| bool can_access_peer(device peer)
+| Queries if this device may directly access the peer device's memory.
+Returns true if direct access is supported.
+
+| void enable_peer_access(device peer)
+| Enables this device to access USM device allocations located on the peer
+device. This does not permit the peer device to access this device's memory.
+Throws an exception if access cannot be enabled or if access is already
+enabled. This device and the peer device must be in the same context.
```

Comment:
The last sentence should be removed "This device and the peer device must be in the same context".  I want to clarify the relationship with context in the core SYCL spec.  In any case, the statement here is unclear.  What context are you talking about?

Reply:
The same context as the allocations being accessed.

Reply:
Which allocations?  This is a *device* query, and there could be many different kernels submitted to this device.  Those kernels might have USM allocations from several different contexts.